Source/Dafny/DafnyAst.cs

```diff
@@ -4751,7 +4798,7 @@ private TupleTypeDecl(ModuleDefinition systemModule, List<TypeParameter> typeArg
 
     public override string CompileName {
       get {
-        return "Tuple" + Dims;
+        return BuiltIns.TupleTypeName(GhostArguments).Replace("_tuple#", "Tuple");
```

Comment:
I think for the runtime name, only the non-ghost arguments matter, so the type would be `"Tuple" + GhostArguments.Count(x => !x)`

Reply:
Correct. And furthermore, I think we should work on the compilers so that `_tuple#1` is never needed.

Reply:
Changing the implementation to `"Tuple" + NonGhostDims` causes the errors reported below when running `dafny /compile:0` on the following program.

```
// RUN: %dafny /compile:0 /dprint:- "%s" /env:0 > "%t"
// RUN: %diff "%s.expect" "%t"

method Test()
{
    // Tuples with a single ghost component are allowed and they are compiled to empty tuples.
    var x := (ghost 123);
    var (y) := (ghost 123);
    ghost var z: (ghost int) := (ghost 123);
}
```

```
(0,-1): Error: more than one declaration of function name: Tclass._System.Tuple0
(0,-1): Error: more than one declaration of variable name: Tagclass._System.Tuple0
(0,-1): Error: more than one declaration of function name: $IsA#_System.Tuple0
(0,-1): Error: more than one declaration of function name: _System.Tuple0#Equal
(0,-1): Error: more than one declaration of variable name: class._System.Tuple0
5 name resolution errors detected in /var/folders/p3/fp88646n1_d92mg86l0knmz00000gs/T/AllowedGhost__module.bpl

*** Encountered internal translation error - re-running Boogie to get better debug information

/var/folders/p3/fp88646n1_d92mg86l0knmz00000gs/T/AllowedGhost__module.bpl(2829,9): Error: more than one declaration of function name: Tclass._System.Tuple0
/var/folders/p3/fp88646n1_d92mg86l0knmz00000gs/T/AllowedGhost__module.bpl(2831,0): Error: more than one declaration of variable name: Tagclass._System.Tuple0
/var/folders/p3/fp88646n1_d92mg86l0knmz00000gs/T/AllowedGhost__module.bpl(2857,9): Error: more than one declaration of function name: $IsA#_System.Tuple0
/var/folders/p3/fp88646n1_d92mg86l0knmz00000gs/T/AllowedGhost__module.bpl(2870,9): Error: more than one declaration of function name: _System.Tuple0#Equal
/var/folders/p3/fp88646n1_d92mg86l0knmz00000gs/T/AllowedGhost__module.bpl(2882,0): Error: more than one declaration of variable name: class._System.Tuple0
5 name resolution errors detected in /var/folders/p3/fp88646n1_d92mg86l0knmz00000gs/T/AllowedGhost__module.bpl
```

It seems that each `TupleTypeDecl` generates its own Boogie variables  and functions based on `CompileName`.

Reply:
I see. That's a bug. That is, `Translator.cs` should not use `.CompileName`. The Boogie names having to do with 2-tuples should look something like `#_System._tuple#2`, whereas the compile names may look like `_System.Tuple2`. It seems the verifier currently uses a mix of these names, which is not right.

I think it will be easiest for the verifier to use one different name for each type. That is, the tuple types `(,)` and `(ghost,)` would in Boogie have names like `_tuple#2oo` and `..._tuple#2go`. There will also be a Boogie name for the 1-tuple `(ghost)`.

The compiler can use names like `"Tuple" + NonGhostDims`. Eventually, we should optimize away `Tuple1`, but that can be a separate pull request.

Reply:
I should have fixed the bug. There was a single call to `ctor.CompileName` in `Resolver.cs` that I changed to `ctor.Name`.